File: 223. Rectangle Area.cpp
class Solution {
public:
    int computeArea(int A, int B, int C, int D, int E, int F, int G, int H) {
        int x1, y1, x2, y2;
        x1 = max(A, E);
        y1 = max(B, F);
        x2 = min(C, G);
        y2 = min(D, H);
        x2 = max(x1, x2);
        y2 = max(y1, y2);
        return (C - A) * (D - B) + (G - E) * (H - F) - (x2 - x1) * (y2 - y1);
    }
};


class Solution {
public:
	int computeArea(int A, int B, int C, int D, int E, int F, int G, int H) {
		long x = 0, y = 0;
		if (A > G || C < E) { // 计算X轴的重叠区域
			
		}
		else {
			x = min(C, G) - max(A, E);
		}

		if (B > H || D < F) // 计算y轴的重叠区域
		{

		}
		else {
			y = min(D, H) - max(B, F);
		}

		return long(C - A)*long(D - B) + long(G - E)*long(H - F) - x * y;
	}
};

File: go/src/leetcode/223. Rectangle Area/223. Rectangle Area.go
/**
@author ZhengHao Lou
Date    2021/09/30

https://leetcode-cn.com/problems/rectangle-area/

*/
package main

import "fmt"

func computeArea(ax1 int, ay1 int, ax2 int, ay2 int, bx1 int, by1 int, bx2 int, by2 int) int {
	var s = (ax2 - ax1) * (ay2 - ay1) + (bx2 - bx1) * (by2 - by1)
	if bx1 >= ax2 || ax1 >= bx2 {
		return s
	}
	if by1 >= ay2 || ay1 >= by2 {
		return s
	}

	var cx1, cy1, cx2, cy2 int
	// left-down
	cx1 = max(ax1, bx1)
	cy1 = max(ay1, by1)
	// right-up
	cx2 = min(bx2, ax2)
	cy2 = min(by2, ay2)
	return s - (cx2 - cx1) * (cy2 - cy1)
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	fmt.Println(computeArea(-3,0,3,4,0,-1,9,2))
}